server/http_handler.go

```diff
@@ -70,26 +71,223 @@ type kvStore interface {
 	SendReq(bo *tikv.Backoffer, req *tikvrpc.Request, regionID tikv.RegionVerID, timeout time.Duration) (*tikvrpc.Response, error)
 }
 
-type regionHandlerTool struct {
+func writeError(w http.ResponseWriter, err error) {
+	w.WriteHeader(http.StatusBadRequest)
+	_, err = w.Write([]byte(err.Error()))
+	terror.Log(errors.Trace(err))
+}
+
+func writeData(w http.ResponseWriter, data interface{}) {
+	js, err := json.Marshal(data)
+	if err != nil {
+		writeError(w, err)
+		return
+	}
+	log.Info(string(js))
+	// write response
+	w.Header().Set(headerContentType, contentTypeJSON)
+	w.WriteHeader(http.StatusOK)
+	_, err = w.Write(js)
+	terror.Log(errors.Trace(err))
+}
+
+type httpHandlerTool struct {
 	bo          *tikv.Backoffer
 	regionCache *tikv.RegionCache
 	store       kvStore
 }
 
-// RegionHandler is the common field for http region handler. It contains
-// some common functions for all handlers.
-type regionHandler struct {
-	*regionHandlerTool
+// newHTTPHandlerTool checks and prepares for region handler.
+// It would panic when any error happens.
+func (s *Server) newHTTPHandlerTool() *httpHandlerTool {
+	var tikvStore kvStore
+	store, ok := s.driver.(*TiDBDriver)
+	if !ok {
+		panic("Invalid KvStore with illegal driver")
+	}
+
+	if tikvStore, ok = store.store.(kvStore); !ok {
```

Comment:
I define `httpHandlerTool` as `regionHandler` before since region handler always needs a `regionCache`. While for the request like `schema` or `settings`, I think they do not need it and maybe their store may not `tikv`. So I do not think it is a good idea to rename `regionHandler` to `httpHandlerTool` directly.

Reply:
I think the elements in the base struct could not be used in all derive struct. It is acceptable.